### PR TITLE
Sticky perf

### PR DESCRIPTION
**Pull request by tharshada**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ npm run change`

#### Description of changes

1) Setting initial scrollbar height and width equal to 0 to prevent rendering when it's width/height changes from undefined to 0.
2) Cascading updates caused due to notifySubscribers for sticky componentDidMount lifecycle hook: addSticky is called from sticky componentDidMount
image
image

3) Reducing computations for sticky component by combining updates for change in sticky state (top/bottom/non-sticky) & distanceFromTop.
4) Setting placeHolder width equal to scrollWidth for nonStickyContent.firstElementChild.

#8430 #7545
#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8542)

### CODE REVIEW

**Comment by tharshada** [comment]
@shagra-ms @leddie24 @ThomasMichon @KevinTCoughlin please review.

**Comment by betrue-final-final** [comment]
Can we get rid of this extra margin on the left?
image

**Comment by shagra-ms** [comment]
>   public distanceFromTop: number;

Please remove this, since we have this in state now.

---
Refers to: packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx:29 in e86b55e. [](commit_id = e86b55e7c91bc4ef1f96c610be381f47d28abea5, deletion_comment = False)

File: packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx [hunk 1]
```diff
@@ -216,9 +214,11 @@ export class Sticky extends BaseComponent<IStickyProps, IStickyState> {
     if (isStickyTop || isStickyBottom) {
       let height = 0,
         width = 0;
-      if (this.nonStickyContent) {
+      if (this.nonStickyContent && this.nonStickyContent.firstElementChild) {
         height = this.nonStickyContent.offsetHeight;
-        width = this.nonStickyContent.scrollWidth;
+        width =
+          this.nonStickyContent.firstElementChild.scrollWidth +
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx]: Please add a code comment on why this is done, and what are the assumptions.

Comment (tharshada) [reply packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx]: sure

**Comment by tharshada** [comment]
> > public distanceFromTop: number;
> 
> Please remove this, since we have this in state now.
> 
> Refers to: packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx:29 in [e86b55e](https://github.com/OfficeDev/office-ui-fabric-react/commit/e86b55e7c91bc4ef1f96c610be381f47d28abea5). [](commit_id = [e86b55e](https://github.com/OfficeDev/office-ui-fabric-react/commit/e86b55e7c91bc4ef1f96c610be381f47d28abea5), deletion_comment = False)

ok

**Comment by shagra-ms** [comment]
@ThomasMichon Can you please take a look?

**Comment by tharshada** [comment]
> Looks good
> 
> 1. Can we address @shagra-ms's comment on removing the `distanceFromTop` property and keeping it in component state only?
> 2. Needs a comment on line 221 of `Sticky.tsx` still for @shagra-ms's comment.
> 3. @betrue-final-final @tharshada , regarding the extra margin on the left for screener tests:  What do should we do to fix this? I think we need a story to account for @tharshada's scenario.  Can we revert the GroupedList example back to the original state (without padding) and just add a new story?
> 4. I think there are additional perf improvements that we can make to the component (e.g: Reducing number of calls to `element.offsetTop` and storing that in the Sticky's state instead), but we can visit that on another PR.
> 5. @tharshada @shagra-ms, does this PR fix #7853 then?  Should I close #8430 once this is merged?

 1.  @leddie24 @shagra-ms  It is actually a good ask but if we see how setState() works, there could be some complications while sorting stickies. Reading distanceFromTop from state may not always give the updated value as state updates may be async. React may combine multiple state updates. However, class variable distanceFromTop is always updated (synchronously). Even if logic for sorting is moved to componentDidUpate(), it won't work because it is possible states (distanceFromTop) of other sticky components are not updated. We are only inserting a given sticky using sorting logic (and not re-sorting all the stickies). For a given operation (Scroll/PLT/etc.), when sticky componentDidUpdate() is called for the last time, it means all states are updated but we've sort all the stickies, which is expensive as we can't find which componentDidUpdate() call is the last one. :(
​​​​​​​
```
 private _setDistanceFromTop(distance: number): void {
    this.setState({ distanceFromTop: distance });
  }
```
```
public shouldComponentUpdate(nextProps: IStickyProps, nextState: IStickyState): boolean {
    ...
    return (... || this.state.distanceFromTop !== nextState.distanceFromTop || ...) as boolean;
  }
```
```
  private _onScrollEvent = (container: HTMLElement, footerStickyContainer: HTMLElement): void => {
    if (this.root && this.nonStickyContent) {
      const distanceFromTop = this._getNonStickyDistanceFromTop(container);
      ... 
      this.setState({
        isStickyTop: this.canStickyTop && isStickyTop,
        isStickyBottom: isStickyBottom,
        distanceFromTop: distanceFromTop
      });
   }
 }
```
```
  public componentDidUpdate(prevProps: IStickyProps, prevState: IStickyState): void {
    const { scrollablePane } = this.context;
    let syncScroll: boolean = false;
    if (!scrollablePane) {
      return;
    }
    const { isStickyTop, isStickyBottom, distanceFromTop } = this.state;

    if (distanceFromTop !== prevState.distanceFromTop) {
      // sort sticky in it's respective container
      scrollablePane.sortSticky(this, true);
      syncScroll = true;
    }

    if (prevState.isStickyTop !== isStickyTop || prevState.isStickyBottom !== isStickyBottom) {
      if (this._activeElement) {
        this._activeElement.focus();
      }
      scrollablePane.updateStickyRefHeights();
      syncScroll = true;
    }

    if (syncScroll) {
      // Sync Sticky scroll position with content container on each update
      scrollablePane.syncScrollSticky(this);
    }
  }
```

But I am trying to find a way to handle this without contributing to complexity.

 2.  @leddie24  I will add the comment.
 3.  @betrue-final-final @leddie24, I have added this test ['ScrollablePane Grouped Details List'](https://github.com/OfficeDev/office-ui-fabric-react/pull/7545/files) for handling horizontal scrollbar issue. The updates are intentional.
 4.  @leddie24  I am working on finding areas to improve perf for sticky. There 'll be upcoming PRs. Suggestions are welcome.
 5.  @leddie24 , I don't think issue #7853 can be handled completely.  Let it remain **open**, keep trying to find the best solution. 
I think we've to come up with some assumptions/best practices, like:
Best practices:
            1. Content inside `<Sticky>` should always be wrapped in a single div.
                `<Sticky><div id={'firstElementChild'}>{intended_content}</div> <Sticky/>`
This can be done in the render() method or by the user.
            2. -ve padding, margin, etc. should not be used.
Assumption:
          1. scrollWidth of a parent is greater than or equal to max of scrollWidths of it's children and same holds for children.

     placeHolder width should be computed in the best possible way to prevent overscroll/underscroll

**Comment by tharshada** [comment]
> Can we get rid of this extra margin on the left?

@betrue-final-final The vr tests are updated on purpose. I will double check the new/changed states once screener task is done.

File: packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx [hunk 2]
```diff
@@ -186,14 +194,7 @@ export class Sticky extends BaseComponent<IStickyProps, IStickyState> {
   }
 
   private _setDistanceFromTop(distance: number): void {
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx]: >_setDistanceFromTop [](start = 10, length = 19)

We are mostly setting 'distanceFromTop' directly in state at other places as well now. We can remove this. There is only once consumer for this anyways.

**Comment by shagra-ms** [comment]
@leddie24 Can you please take a look again? The changes look fine to me now.

**Comment by shagra-ms** [comment]
@betrue-final-final Can you take a look at the failed screener tests? @tharshada has clarified the reason for extra padding as well.

**Comment by khmakoto** [comment]
@betrue-final-final could you take a look at the failed screener tests? The changes seem to be intentional.

File: packages/file-type-icons/fileTypeIconMap.json [hunk 3]
```diff
@@ -510,4 +510,4 @@
   "zip": {
     "extensions": ["zip"]
   }
-}
\ No newline at end of file
+}
```
Comment (khmakoto) [inline packages/file-type-icons/fileTypeIconMap.json]: This new line was added after fixing the conflict.

Comment (shagra-ms) [reply packages/file-type-icons/fileTypeIconMap.json]: @tharshada can you please resolve this?

File: packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx [hunk 4]
```diff
@@ -216,9 +214,28 @@ export class Sticky extends BaseComponent<IStickyProps, IStickyState> {
...
+      // It is necessary to provide a placeHolder of a certain width (height is already being set) in the content--container,
+      // to get a horizontal scrollbar & be able to view the complete content of sticky component.
+      if (this.nonStickyContent && this.nonStickyContent.firstElementChild) {
         height = this.nonStickyContent.offsetHeight;
-        width = this.nonStickyContent.scrollWidth;
+        // What value should be substituted for placeHolder width?
+        // Assumption:
+        //    1. Content inside <Sticky> should always be wrapped in a single div.
+        //        <Sticky><div id={'firstElementChild'}>{intended_content}</div><Sticky/>
+        //    2. -ve padding, margin, etc. are not be used.
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx]: What is "-ve padding"?

File: packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.base.tsx [hunk 5]
```diff
@@ -363,7 +362,7 @@ export class ScrollablePaneBase extends BaseComponent<IScrollablePaneProps, IScr
         // Get first element that has a distance from top that is further than our sticky that is being added
         let targetStickyToAppendBefore: Sticky | undefined = undefined;
         for (const i in stickyListSorted) {
-          if (stickyListSorted[i].distanceFromTop >= sticky.distanceFromTop) {
+          if ((stickyListSorted[i].state.distanceFromTop || 0) >= (sticky.state.distanceFromTop || 0)) {
```
[the remark this answers is not in the document]
Comment (tharshada) [reply packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.base.tsx]: image
ScrollablePane.Default.Example.tsx order is affected.
image

**Comment by tharshada** [comment]
> Hey @tharshada, I had difficulty confirming the performance benefits locally using the profiler.
> 
> > 1. Setting initial scrollbar height and width equal to 0 to prevent rendering when it's width/height changes from undefined to 0.
> 
> Your code changes make sense. However, I was unable to see in the profiler that rendering is prevented.
> 
> Can you snapshot a profiler capture showing the above savings?
> 
> > 1. Cascading updates caused due to notifySubscribers for sticky componentDidMount lifecycle hook: addSticky is called from sticky componentDidMount
> 
> In both `master` and this branch, I see roughly ~5x cascading updates in Chromium's "User Timings" pane. Here's one instance with your changes:
> 
> image
> 
> Overall, its not clear to me what performance implications this PR has looking at the profiler snapshots myself. It would be _really_ helpful if you could share before / after profiler comparisons.
> 
> I'm also happy to have a call this week if that's easier.
> 
> Addressing performance with `Sticky` / `ScrollablePane` will benefit document libraries and Fabric greatly, so kudos for the effort here. I outlined some of the problems we've seen in [#8780 (comment)]

Cascading updates for Sticky.componentDidMount() are caused due to 
1) sticky.setDistanceFromTop(this.contentContainer); this causes a forceUpdate() if distanceFromTop has changed.

image

2) this.notifySubscribers(); this calls setState() for all other mounted Sticky components including for itself. For Lists/DocLibs, there are 3 sticky components- Title/Breadcrumb, DetailsHeader, DetailsFooter. For every sticky component, setState() is called N times when a sticky is in componentDidMount() lifecyle, where N is the number of mounted sticky components at that time. 
image
        notifySubscribers() has been removed for sticky.componentDidMount(). This stack would go away.

**Comment by KevinTCoughlin** [comment]
@tharshada here's what I see with `ScrollablePaneDetailsListExample` on initial render:

### URL & Code Used

http://localhost:4322/#/tests/detailslistbasicexample

```diff
diff --git a/apps/fabric-website-resources/src/AppDefinition.tsx b/apps/fabric-website-resources/src/AppDefinition.tsx
index dfa3e2436..01ca5c6c1 100644
--- a/apps/fabric-website-resources/src/AppDefinition.tsx
+++ b/apps/fabric-website-resources/src/AppDefinition.tsx
@@ -1,6 +1,7 @@
 import * as React from 'react';
 import { App as AppBase, IAppDefinition, IAppProps } from '@uifabric/example-app-base';
-import { DetailsListBasicExample } from 'office-ui-fabric-react/lib/components/DetailsList/examples/DetailsList.Basic.Example';
+// import { DetailsListBasicExample } from 'office-ui-fabric-react/lib/components/DetailsList/examples/DetailsList.Basic.Example';
+import { ScrollablePaneDetailsListExample } from 'office-ui-fabric-react/lib/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example';
 import { AppCustomizations } from './customizations/customizations';
 
 export const AppDefinition: IAppDefinition = {
@@ -8,9 +9,9 @@ export const AppDefinition: IAppDefinition = {
   customizations: AppCustomizations,
   testPages: [
     {
-      component: DetailsListBasicExample,
-      key: 'DetailsListBasicExample',
-      name: 'DetailsListBasicExample',
+      component: ScrollablePaneDetailsListExample,
+      key: 'ScrollablePaneDetailsListExample',
+      name: 'ScrollablePaneDetailsListExample',
       url: '#/tests/detailslistbasicexample'
     }
   ],
```

### Before

1. image
1. image
1. image

### After

1. image
1. image
1. image

It seems like we eliminated `3)` which contained the reflow, is that correct? If so 🎉🎉🎉

**Comment by tharshada** [comment]
> @tharshada here's what I see with ScrollablePaneDetailsListExample on initial render:
> URL & Code Used
> http://localhost:4322/#/tests/detailslistbasicexample
> diff --git a/apps/fabric-website-resources/src/AppDefinition.tsx b/apps/fabric-website-resources/src/AppDefinition.tsx
> index dfa3e2436..01ca5c6c1 100644
> --- a/apps/fabric-website-resources/src/AppDefinition.tsx
> +++ b/apps/fabric-website-resources/src/AppDefinition.tsx
> @@ -1,6 +1,7 @@
>  import * as React from 'react';
>  import { App as AppBase, IAppDefinition, IAppProps } from '@uifabric/example-app-base';
> -import { DetailsListBasicExample } from 'office-ui-fabric-react/lib/components/DetailsList/examples/DetailsList.Basic.Example';
> +// import { DetailsListBasicExample } from 'office-ui-fabric-react/lib/components/DetailsList/examples/DetailsList.Basic.Example';
> +import { ScrollablePaneDetailsListExample } from 'office-ui-fabric-react/lib/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example';
>  import { AppCustomizations } from './customizations/customizations';
>  
>  export const AppDefinition: IAppDefinition = {
> @@ -8,9 +9,9 @@ export const AppDefinition: IAppDefinition = {
>    customizations: AppCustomizations,
>    testPages: [
>      {
> -      component: DetailsListBasicExample,
> -      key: 'DetailsListBasicExample',
> -      name: 'DetailsListBasicExample',
> +      component: ScrollablePaneDetailsListExample,
> +      key: 'ScrollablePaneDetailsListExample',
> +      name: 'ScrollablePaneDetailsListExample',
>        url: '#/tests/detailslistbasicexample'
>      }
>    ],
> Before
> 
> 
> 
> 
> 
> After
> 
> 
> 
> 
> 
> It seems like we eliminated 2 & 3 which contained the reflow, is that correct? If so 🎉🎉🎉.
> Still not confident I'm repro'ing correctly. Steps were apply the above code and capture browser reload. Then grep flame graph for Sticky.componentDidMount and look at flame.


image

this syncScroll() is now being done in componentDidUpdate() if distanceFromTop has changed. It won't be under componentDidMount(). Yes, it was causing reflow here. Sticky._onScroll() calls setState() by combining updates due to change in distanceFromTop & sticky state. Earlier syncScroll() was getting called separately- 1) Sticky.componentDidMount()=> notifySubscriber=>forceUpdate(), 2) in Sticky.ComponentDidUpdate().  For componentDidUpdate(),  syncScoll() 'll get called and reflow 'll still exist. I am working on removing layout calculations due to syncScroll for PLT ('ll be in another PR).
To enhance PLT, optimizing componentDidMount() seemed like a good place to start.

**Comment by KevinTCoughlin** [comment]
@tharshada awesome work! Thanks for the explainer in https://github.com/OfficeDev/office-ui-fabric-react/pull/8542#issuecomment-486743184 and for your patience. Change LGTM.

**Comment by KevinTCoughlin** [comment]
@tharshada, I tried to resolve conflicts, but I lack permission to push to your fork :(. You can simply accept the incoming changes to resolve and we should be good to merge this.

**Pull request merged**